src/clr-addons/datagrid/datagrid-state-persistence/state-persistence-key.directive.ts

```diff
@@ -1,35 +1,156 @@
+      const state = this.getLocalStorageState();
+      state.pageSize = pageSize;
+      localStorage.setItem(this.options.key, JSON.stringify(state));
+    });
+
+    /* init page size of datagrid if already persisted in local storage */
+    if (savedState.pageSize) {
+      this.pagination.page.size = savedState.pageSize;
+    }
+  }
+
+  private initCurrentPage(savedState: ClrDatagridStatePersistenceModel) {
+    /* init current page of datagrid if already persisted */
+    if (savedState?.currentPage) {
+      this.pagination.page.current = savedState.currentPage;
+    }
+  }
+
+  private initFilter(savedState: ClrDatagridStatePersistenceModel) {
+    if (savedState.columns) {
+      Object.keys(savedState.columns).forEach(prop => {
+        const filter = this.getFilter(prop);
+        if (filter) {
+          const savedFilterValue = savedState.columns[prop].filterValue || {};
+          Object.keys(savedFilterValue)
+            .filter(valueKey => valueKey !== 'property')
+            .forEach(valueKey => ((filter as any)[valueKey] = this.parseFilterValue(savedFilterValue[valueKey])));
         }
-        state.pageSize = pageSize;
-        localStorage.setItem(this.clrStatePersistenceKey, JSON.stringify(state));
       });
-
-      /* init page size of datagrid if already persisted in local storage */
-      const loadedState = JSON.parse(localStorage.getItem(this.clrStatePersistenceKey));
-      if (loadedState?.pageSize) {
-        /* postpone set size to other cycle as it is already set in this change detection cycle */
-        setTimeout(() => (this.pagination.page.size = loadedState.pageSize));
-      }
     }
   }
+
+  private initDatagridPersister() {
+    this.datagrid.refresh.pipe(takeUntil(this.destroy$)).subscribe(dgState => {
+      const state = this.getSessionState();
+
+      state.currentPage = dgState.page?.current;
+      state.columns = state.columns || {};
+      Object.keys(state.columns).forEach(prop => (state.columns[prop].filterValue = undefined));
+      dgState.filters?.forEach(filter => {
+        const property = this.getFilterPropertyName(filter);
+        if (property) {
+          state.columns[property] = state.columns[property] || {};
+          state.columns[property].filterValue = this.enrichFilterValue(filter);
+        }
+      });
+
+      sessionStorage.setItem(this.options.key, JSON.stringify(state));
+    });
+  }
+
+  getSessionState(): ClrDatagridStatePersistenceModel {
+    return JSON.parse(sessionStorage.getItem(this.options.key)) || {};
+  }
+
+  getLocalStorageState(): ClrDatagridStatePersistenceModel {
+    return JSON.parse(localStorage.getItem(this.options.key)) || {};
+  }
+
+  /**
+   * As a date is serialized as string, but not deserialized as date
+   * we need to add some meta information to do that manually later
+   */
+  private enrichFilterValue(filter: any) {
+    const result = {} as any;
```

Comment:
we're gonna be returning any here, which is not ideal. Could we use a `Record` type here as the return value? like `Record<keyof filter, unknown>`

Reply:
yes oversaw that, will remove the any

Reply:
done